"vbc.exe" error code -2146232797 when a Derived Class has a Shared method that handles an event from a Shared WithEvents variable in its Base Class

https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1263908

_This issue has been moved from [a ticket on Developer Community](https://developercommunity2.visualstudio.com/t/vbcexe-error-code--2146232797-when-a-/1305265)._

---
[severity:I'm unable to use this version]
Visual Studio produces the following error when compiling an application where a Derived Class has a Shared method that handles an event from a Shared WithEvents variable in its Base Class:

>error MSB6006: "vbc.exe" exited with code -2146232797.

The following is a sample module, in a Visual Basic .NET Framework Console application, that produces the error:

```
Module Module1
    Sub Main()
    End Sub
End Module

Public Class DerivedClass
    Inherits BaseClass

    Private Shared Sub Handler() Handles EventHost.SomeEvent
    End Sub
End Class

Public MustInherit Class BaseClass
    Protected Shared WithEvents EventHost As EventHost
End Class

Public Class EventHost
    Public Event SomeEvent()
End Class
```

---
### Original Comments

#### Feedback Bot on 1/10/2021, 07:25 PM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

#### Feedback Bot on 1/14/2021, 09:08 AM: 

<p>This issue is currently being investigated. Our team will get back to you if either more information is needed, a workaround is available, or the issue is resolved.</p>


---
### Original Solutions
(no solutions)

--- a/src/Compilers/VisualBasic/Portable/Errors/Errors.vb
+++ b/src/Compilers/VisualBasic/Portable/Errors/Errors.vb
@@ -1384,7 +1384,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         'ERR_QueryStartsWithLet = 36608
         'ERR_NoQueryExpressionsInDebugger = 36609
         ERR_QueryNameNotDeclared = 36610
-        '// Available 36611
+
+        ERR_SharedEventNeedsHandlerInTheSameType = 36611
 
         ERR_NestedFunctionArgumentNarrowing3 = 36612
 

--- a/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberMethodSymbol.vb
+++ b/src/Compilers/VisualBasic/Portable/Symbols/Source/SourceMemberMethodSymbol.vb
@@ -663,9 +663,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                     Return Nothing
                 End If
 
-                If witheventsProperty.IsShared AndAlso Not Me.IsShared Then
-                    'Events of shared WithEvents variables cannot be handled by non-shared methods.
-                    Binder.ReportDiagnostic(diagBag, singleHandleClause.EventContainer, ERRID.ERR_SharedEventNeedsSharedHandler)
+                Dim isFromBase = Not TypeSymbol.Equals(witheventsProperty.ContainingType, Me.ContainingType, TypeCompareKind.ConsiderEverything)
+
+                If witheventsProperty.IsShared Then
+                    Debug.Assert(Not witheventsProperty.IsOverridable)
+
+                    If Not Me.IsShared Then
+                        'Events of shared WithEvents variables cannot be handled by non-shared methods.
+                        Binder.ReportDiagnostic(diagBag, singleHandleClause.EventContainer, ERRID.ERR_SharedEventNeedsSharedHandler)
+
+                    End If
+
+                    If isFromBase Then
+                        Binder.ReportDiagnostic(diagBag, singleHandleClause.EventContainer, ERRID.ERR_SharedEventNeedsHandlerInTheSameType)
+                    End If
                 End If
 
                 If eventContainerKind = SyntaxKind.WithEventsPropertyEventContainer Then
@@ -688,7 +699,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
                 End If
 
                 ' if was found in one of bases, need to override it
-                If Not TypeSymbol.Equals(witheventsProperty.ContainingType, Me.ContainingType, TypeCompareKind.ConsiderEverything) Then
+                If isFromBase Then
                     witheventsPropertyInCurrentClass = DirectCast(Me.ContainingType, SourceNamedTypeSymbol).GetOrAddWithEventsOverride(witheventsProperty)
                 Else
                     witheventsPropertyInCurrentClass = witheventsProperty

--- a/src/Compilers/VisualBasic/Portable/VBResources.resx
+++ b/src/Compilers/VisualBasic/Portable/VBResources.resx
@@ -1301,6 +1301,9 @@
   <data name="ERR_SharedEventNeedsSharedHandler" xml:space="preserve">
     <value>Events of shared WithEvents variables cannot be handled by non-shared methods.</value>
   </data>
+  <data name="ERR_SharedEventNeedsHandlerInTheSameType" xml:space="preserve">
+    <value>Events of shared WithEvents variables cannot be handled by methods in a different type.</value>
+  </data>
   <data name="ERR_ExpectedMinus" xml:space="preserve">
     <value>'-' expected.</value>
   </data>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.cs.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.cs.xlf
@@ -37,6 +37,11 @@
         <target state="translated">Cílový modul runtime nepodporuje přístupnost Protected, Protected Friend nebo Private Protected pro člena rozhraní.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">komentáře po pokračování řádku</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.de.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.de.xlf
@@ -37,6 +37,11 @@
         <target state="translated">Die Zielruntime bietet keine Unterstützung für die Zugriffsmöglichkeiten "Protected", "Protected Friend" oder "Private Protected" für einen Member einer Schnittstelle.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">Kommentare nach Zeilenfortsetzung</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.es.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.es.xlf
@@ -37,6 +37,11 @@
         <target state="translated">El entorno de ejecución de destino no admite la accesibilidad de tipo "Protected", "Protected Friend" o "Private Protected" para un miembro de una interfaz.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">comentarios después de la continuación de línea</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.fr.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.fr.xlf
@@ -37,6 +37,11 @@
         <target state="translated">Le runtime cible ne prend pas en charge l'accessibilité 'Protected', 'Protected Friend' ou 'Private Protected' pour un membre d'interface.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">commentaires après la continuation de la ligne</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.it.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.it.xlf
@@ -37,6 +37,11 @@
         <target state="translated">Il runtime di destinazione non supporta l'accessibilità 'Protected', 'Protected Friend' o 'Private Protected' per un membro di un'interfaccia.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">commenti dopo la continuazione di riga</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.ja.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.ja.xlf
@@ -37,6 +37,11 @@
         <target state="translated">ターゲット ランタイムは、インターフェイスのメンバーの 'Protected'、'Protected Friend'、'Private Protected' のアクセシビリティをサポートしていません。</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">行連結後のコメント</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.ko.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.ko.xlf
@@ -37,6 +37,11 @@
         <target state="translated">대상 런타임은 인터페이스의 멤버에 대해 'Protected', 'Protected Friend' 또는 'Private Protected' 접근성을 지원하지 않습니다.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">줄 연속 뒤 주석</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.pl.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.pl.xlf
@@ -37,6 +37,11 @@
         <target state="translated">Docelowe środowisko uruchomieniowe nie obsługuje poziomu dostępu „Protected”, „Protected Friend” ani „Private Protected” dla składowej interfejsu.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">komentarze po kontynuacji wiersza</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.pt-BR.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.pt-BR.xlf
@@ -37,6 +37,11 @@
         <target state="translated">O tempo de execução de destino não dá suporte à acessibilidade 'Protected', 'Protected Friend' ou 'Private Protected' para um membro de uma interface.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">comentários após a continuação da linha</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.ru.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.ru.xlf
@@ -37,6 +37,11 @@
         <target state="translated">Целевая среда выполнения не поддерживает доступ "Protected", "Protected Friend" или "Private Protected" для элемента интерфейса.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">комментарии после продолжения строки</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.tr.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.tr.xlf
@@ -37,6 +37,11 @@
         <target state="translated">Hedef çalışma zamanı, bir arabirimin üyesi için 'Protected', 'Protected Friend' veya 'Private Protected' erişilebilirliğini desteklemez.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">satır devamı sonrası açıklamalar</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.zh-Hans.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.zh-Hans.xlf
@@ -37,6 +37,11 @@
         <target state="translated">目标运行时不支持对接口的成员使用 "Protected"、"Protected Friend" 或 "Private Protected" 辅助功能。</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">行继续符之后的注释</target>

--- a/src/Compilers/VisualBasic/Portable/xlf/VBResources.zh-Hant.xlf
+++ b/src/Compilers/VisualBasic/Portable/xlf/VBResources.zh-Hant.xlf
@@ -37,6 +37,11 @@
         <target state="translated">目標執行階段不支援對介面成員的 'Protected'、'Protected Friend' 或 'Private Protected' 存取範圍。</target>
         <note />
       </trans-unit>
+      <trans-unit id="ERR_SharedEventNeedsHandlerInTheSameType">
+        <source>Events of shared WithEvents variables cannot be handled by methods in a different type.</source>
+        <target state="new">Events of shared WithEvents variables cannot be handled by methods in a different type.</target>
+        <note />
+      </trans-unit>
       <trans-unit id="FEATURE_CommentsAfterLineContinuation">
         <source>comments after line continuation</source>
         <target state="translated">行接續符號後註解</target>

--- a/src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/SymbolErrorTests.vb
+++ b/src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/SymbolErrorTests.vb
@@ -6359,6 +6359,7 @@ BC31029: Method 'EventHandler' cannot handle event 'Event1' because they do not 
                                                                           ~~~~~~
                  ]]></errors>
             CompilationUtils.AssertTheseDeclarationDiagnostics(compilation1, expectedErrors1)
+            CompilationUtils.AssertTheseEmitDiagnostics(compilation1, expectedErrors1)
         End Sub
 
         <Fact>
@@ -22329,6 +22330,149 @@ BC30270: 'Iterator' is not valid on an interface method declaration.
 ]]></errors>
 
             CompilationUtils.AssertTheseDeclarationDiagnostics(compilation, expectedErrors)
+        End Sub
+
+        <Fact>
+        <WorkItem(50472, "https://github.com/dotnet/roslyn/issues/50472")>
+        <WorkItem(1263908, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1263908")>
+        Public Sub UnsupportedHandles_01()
+            Dim compilation = CompilationUtils.CreateCompilation(
+<compilation>
+    <file name="a.vb"><![CDATA[
+Public Class DerivedClass
+    Inherits BaseClass
+
+    Private Shared Sub Handler() Handles EventHost.SomeEvent
+    End Sub
+End Class
+
+Public MustInherit Class BaseClass
+    Protected Shared WithEvents EventHost As EventHost
+End Class
+
+Public Class EventHost
+    Public Event SomeEvent()
+End Class
+    ]]></file>
+</compilation>)
+
+            Dim expectedErrors = <errors><![CDATA[ 
+BC36611: Events of shared WithEvents variables cannot be handled by methods in a different type.
+    Private Shared Sub Handler() Handles EventHost.SomeEvent
+                                         ~~~~~~~~~
+]]></errors>
+
+            compilation.AssertTheseDeclarationDiagnostics(expectedErrors)
+            compilation.AssertTheseEmitDiagnostics(expectedErrors)
+        End Sub
+
+        <Fact>
+        <WorkItem(50472, "https://github.com/dotnet/roslyn/issues/50472")>
+        <WorkItem(1263908, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1263908")>
+        Public Sub UnsupportedHandles_02()
+            Dim compilation = CompilationUtils.CreateCompilation(
+<compilation>
+    <file name="a.vb"><![CDATA[
+Public Class DerivedClass
+    Inherits BaseClass
+
+    Private Sub Handler() Handles EventHost.SomeEvent
+    End Sub
+End Class
+
+Public MustInherit Class BaseClass
+    Protected Shared WithEvents EventHost As EventHost
+End Class
+
+Public Class EventHost
+    Public Event SomeEvent()
+End Class
+    ]]></file>
+</compilation>)
+
+            Dim expectedErrors = <errors><![CDATA[ 
+BC30594: Events of shared WithEvents variables cannot be handled by non-shared methods.
+    Private Sub Handler() Handles EventHost.SomeEvent
+                                  ~~~~~~~~~
+BC36611: Events of shared WithEvents variables cannot be handled by methods in a different type.
+    Private Sub Handler() Handles EventHost.SomeEvent
+                                  ~~~~~~~~~
+]]></errors>
+
+            compilation.AssertTheseDeclarationDiagnostics(expectedErrors)
+            compilation.AssertTheseEmitDiagnostics(expectedErrors)
+        End Sub
+
+        <Fact>
+        <WorkItem(50472, "https://github.com/dotnet/roslyn/issues/50472")>
+        <WorkItem(1263908, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1263908")>
+        Public Sub UnsupportedHandles_03()
+            Dim compilation = CompilationUtils.CreateCompilation(
+<compilation>
+    <file name="a.vb"><![CDATA[
+Public Class DerivedClass
+    Inherits BaseClass
+
+    Private Shared Sub Handler() Handles EventHost.SomeEvent, EventHost2.SomeEvent
+    End Sub
+
+    Protected Shared WithEvents EventHost2 As EventHost
+End Class
+
+Public MustInherit Class BaseClass
+    Protected Shared WithEvents EventHost As EventHost
+End Class
+
+Public Class EventHost
+    Public Event SomeEvent()
+End Class
+    ]]></file>
+</compilation>)
+
+            Dim expectedErrors = <errors><![CDATA[ 
+BC36611: Events of shared WithEvents variables cannot be handled by methods in a different type.
+    Private Shared Sub Handler() Handles EventHost.SomeEvent, EventHost2.SomeEvent
+                                         ~~~~~~~~~
+]]></errors>
+
+            compilation.AssertTheseDeclarationDiagnostics(expectedErrors)
+            compilation.AssertTheseEmitDiagnostics(expectedErrors)
+        End Sub
+
+        <Fact>
+        <WorkItem(50472, "https://github.com/dotnet/roslyn/issues/50472")>
+        <WorkItem(1263908, "https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1263908")>
+        Public Sub UnsupportedHandles_04()
+            Dim compilation = CompilationUtils.CreateCompilation(
+<compilation>
+    <file name="a.vb"><![CDATA[
+Public Class DerivedClass
+    Inherits BaseClass
+
+    Private Shared Sub Handler() Handles EventHost2.SomeEvent, EventHost.SomeEvent
+    End Sub
+
+    Protected Shared WithEvents EventHost2 As EventHost
+End Class
+
+Public MustInherit Class BaseClass
+    Protected Shared WithEvents EventHost As EventHost
+End Class
+
+Public Class EventHost
+    Public Event SomeEvent()
+End Class
+    ]]></file>
+</compilation>)
+
+            Dim expectedErrors = <errors><![CDATA[ 
+BC36611: Events of shared WithEvents variables cannot be handled by methods in a different type.
+    Private Shared Sub Handler() Handles EventHost2.SomeEvent, EventHost.SomeEvent
+                                                               ~~~~~~~~~
+]]></errors>
+
+            compilation.AssertTheseDeclarationDiagnostics(expectedErrors)
+            compilation.AssertTheseEmitDiagnostics(expectedErrors)
         End Sub
 
         <Fact>

